trinity/tx_pool/validators.py

```diff
@@ -0,0 +1,62 @@
+from typing import (
+    Type
+)
+
+from evm.chains.base import (
+    BaseChain
+)
+from evm.exceptions import (
+    ValidationError
+)
+from evm.rlp.transactions import (
+    BaseTransaction,
+    BaseTransactionFields
+)
+
+
+class DefaultTransactionValidator():
+    """
+    The :class:`~trinity.tx_pool.validators.DefaultTransactionValidator` class is responsible to
+    decide wether transactions should be relayed to peers or not. This implementation validates
+    transactions against a transaction class inferred from a ``initial_tx_validation_block_number``
+    but will switch to a different one as soon as the tip of the chain uses a more up to date
+    transaction class than the one that corresponds to the ``initial_tx_validation_block_number``.
+    """
+
+    def __init__(self, chain: BaseChain, initial_tx_validation_block_number: int) -> None:
+        self.chain = chain
+        self._initial_tx_class = self._get_tx_class_for_block_number(
+            initial_tx_validation_block_number
+        )
+        self._ordered_tx_classes = [
+            vm_class.get_transaction_class()
+            for _, vm_class in chain.get_vm_configuration()
+        ]
+        self._initial_tx_class_index = self._ordered_tx_classes.index(self._initial_tx_class)
+
+    def __call__(self, transaction: BaseTransactionFields) -> bool:
+
+        transaction_class = self.get_appropriate_tx_class()
+        tx = transaction_class(**transaction.as_dict())
+        try:
+            tx.validate()
+        except ValidationError:
+            return False
+        else:
+            return True
+
+    def get_appropriate_tx_class(self) -> Type[BaseTransaction]:
+        current_tx_class = self.chain.get_vm_class().get_transaction_class()
+        # If the current head of the chain is still on a fork that is before the currently
+        # active fork (syncing), ensure that we use the specified initial tx class
+        if self._is_outdated_tx_class(current_tx_class):
+            return self._initial_tx_class
+
+        return current_tx_class
+
+    def _is_outdated_tx_class(self, tx_class: Type[BaseTransaction]) -> bool:
+        return self._ordered_tx_classes.index(tx_class) < self._initial_tx_class_index
+
+    def _get_tx_class_for_block_number(self, block_number: int) -> Type[BaseTransaction]:
```

Comment:
I question the necessity of this method over just inlining `self.chain.get_vm_class_for_block_number(block_number).get_transaction_class()` though they both admittedly are not great.  Probably fine either way.

Reply:
I agree that in this particular case it's not too much of an improvement over the inlined version. I often do these things for improved readability. Anyway, I'm going to leave this as is now to be able to merge without another CI run.